include/xtensor/xrandom.hpp

```diff
@@ -0,0 +1,195 @@
+ */
+
+#ifndef XRANDOM_HPP
+#define XRANDOM_HPP
+
+#include <utility>
+#include <random>
+#include <memory>
+#include <sstream>
+
+#include "xgenerator.hpp"
+
+namespace xt
+{
+    namespace detail
+    {
+        template <class D, class S, class V, class E = std::mt19937>
+        struct random_impl
+        {
+            using value_type = V;
+            using size_type = std::size_t;
+
+            random_impl(E engine, D dist, S shape) :
+                m_engine(engine), m_dist(dist), m_strides(shape.size()), 
+                m_engine_state(std::make_shared<std::stringstream>())
+            {
+                std::size_t data_size = 1;
+                for(size_type i = m_strides.size(); i != 0; --i)
+                {
+                    m_strides[i - 1] = data_size;
+                    data_size = m_strides[i - 1] * shape[i - 1];
+                }
+                // save initial engine state
+                (*m_engine_state) << m_engine; 
+            }
+
+            template <class... Args>
+            inline value_type operator()(Args... args) const
+            {
+                std::array<size_type, sizeof...(Args)> args_arr({static_cast<size_type>(args)...});
+                return access_impl(args_arr);
+            }
+
+            inline value_type operator[](const xindex& idx) const
+            {
+                return access_impl(idx);
+            }
+
+            template <class It>
+            inline value_type element(It first, It /*last*/) const
+            {
+                return access_impl(*first);
+            }
+
+        private:
+            mutable E m_engine;
+            mutable D m_dist;
+            mutable S m_strides;
+            // std::stringstream is not copyable but has to be copied in
+            // case of being used in an xfunction
+            std::shared_ptr<std::stringstream> m_engine_state;
+            mutable long m_advance_state = -1;
+
+            template <class A>
+            value_type access_impl(const A& idx) const
+            {
+                long advance_state = std::inner_product(idx.begin(), idx.end(), m_strides.begin(), 0);
+                m_advance_state++;
+                if (m_advance_state == advance_state)
+                {
+                    return m_dist(m_engine);
+                }
+                else
+                {
+                    if (advance_state > m_advance_state)
+                    {
+                        m_engine.discard(advance_state - m_advance_state);
+                        m_advance_state = advance_state;
+                        return m_dist(m_engine);
+                    }
+                    // TODO(BUG): this doesn't reset correctly after 2nd iteration...
+                    (*m_engine_state) >> m_engine;
+                    m_engine.discard(advance_state);
+                    m_advance_state = advance_state;
+
+                    return m_dist(m_engine);
+                }
+            }
+        };
```

Comment:
Hmm, so my basic thinking was that a random number generator puts out a "fixed" sequence of random numbers. 
And then one could go reset and advance the RNG as one pleases, basically the same as having linearly laid out data in a vector. That's what I was using the strides for, to find the correct "advance" position of the RNG. 
And that part is working fine.

Unfortunately, distributions can also carry an internal state (that can be resetted as well). But, afaik, distributions don't offer an `advance` operation. So the `()` operator would have to be called `advance` times in order to arrive at the correct random number. 

This will make random access that doesn't linearly count upwards kinda slow, as there will be a bunch of reset/advance ... cycles.

Is there a better option?